boto3/crt.py

```diff
@@ -0,0 +1,161 @@
+        'region': region_name,
+        'use_ssl': True,
+        'crt_credentials_provider': cred_provider,
+    }
+    return create_s3_crt_client(**create_crt_client_kwargs)
+
+
+def _create_crt_request_serializer(session, region_name):
+    return BotocoreCRTRequestSerializer(
+        session, {'region_name': region_name, 'endpoint_url': None}
+    )
+
+
+def _create_crt_s3_client(session, config, region_name, credentials, **kwargs):
+    """Create boto3 wrapper class to manage crt lock reference and S3 client."""
+    lock = acquire_crt_s3_process_lock(PROCESS_LOCK_NAME)
+    if lock is None:
+        # If we're unable to acquire the lock, we cannot
+        # use the CRT in this process and should default to
+        # the classic s3transfer manager.
+        return None
+
+    cred_wrapper = BotocoreCRTCredentialsWrapper(credentials)
+    cred_provider = cred_wrapper.to_crt_credentials_provider()
+    return CRTS3Client(
+        _create_crt_client(session, config, region_name, cred_provider),
+        lock,
+        region_name,
+        cred_wrapper,
+    )
+
+
+def _initialize_crt_transfer_primatives(client, config):
+    session = Session()
+    region_name = client.meta.region_name
+    credentials = client._get_credentials()
+
+    serializer = _create_crt_request_serializer(session, region_name)
+    s3_client = _create_crt_s3_client(
+        session, config, region_name, credentials
+    )
+    return serializer, s3_client
+
+
+def get_crt_s3_client(client, config):
+    global _CRT_S3_CLIENT
+    global _BOTOCORE_CRT_SERIALIZER
+
+    with CLIENT_CREATION_LOCK:
+        if _CRT_S3_CLIENT is None:
+            serializer, s3_client = _initialize_crt_transfer_primatives(
+                client, config
+            )
+            _BOTOCORE_CRT_SERIALIZER = serializer
+            _CRT_S3_CLIENT = s3_client
+
+    return _CRT_S3_CLIENT
+
+
+class CRTS3Client:
+    """
+    This wrapper keeps track of our underlying CRT client, the lock used to
+    acquire it and the region we've used to instantiate the client.
+
+    Due to limitations in the existing CRT interfaces, we can only make calls
+    in a single region and does not support redirects. We track the region to
+    ensure we don't use the CRT client when a successful request cannot be made.
+    """
+
+    def __init__(self, crt_client, process_lock, region, cred_provider):
+        self.crt_client = crt_client
+        self.process_lock = process_lock
+        self.region = region
+        self.cred_provider = cred_provider
+
+
+def is_crt_compatible_request(client, crt_s3_client):
+    """
+    Boto3 client must use same signing region and credentials
+    as the _CRT_S3_CLIENT singleton. Otherwise fallback to classic.
+    """
+    if crt_s3_client is None:
+        return False
+
+    is_same_region = client.meta.region_name == crt_s3_client.region
+    is_same_identity = compare_identity(
+        client._get_credentials(), crt_s3_client.cred_provider
```

Comment:
A couple of things here:
1. Is it possible for `_get_credentials()` to be `None` here? I'm wondering if we can run into a case where we try to access properties such as `access_key`, `secret_key`, etc on a `None` object.
2. I'm wondering if we should be calling `get_frozen_credentials()` on the return value from `_get_credentials()`? This would guarantee we don't run into race conditions where we access one of the properties and a refresh happens on us in between comparison of another property.